warehouse/query-core/src/main/java/datawave/query/jexl/visitors/UniqueExpressionTermsVisitor.java

```diff
@@ -0,0 +1,199 @@
+                break;
+            }
+            
+            visitor.iterations++;
+        }
+        
+        // Log if we stopped visiting due to exceeding the max iterations
+        if (visitor.iterations >= visitor.maxIterations && visitor.iterations > 0) {
+            if (log.isDebugEnabled()) {
+                log.debug("UniqueExpressionTermsVisitor exceeded maximum iteration (" + visitor.maxIterations + ").");
+            }
+        }
+        
+        if (log.isDebugEnabled()) {
+            log.debug("UniqueExpressionTermsVisitor took " + visitor.iterations + " iterations to fully process query tree.");
+            log.debug("UniqueExpressionTermsVisitor removed " + visitor.duplicates + " duplicate terms");
+        }
+        return (T) copy;
+    }
+    
+    @Override
+    public Object visit(ASTOrNode node, Object data) {
+        removeDuplicateTerms(node);
+        node.childrenAccept(this, data);
+        return data;
+    }
+    
+    @Override
+    public Object visit(ASTAndNode node, Object data) {
+        removeDuplicateTerms(node);
+        node.childrenAccept(this, data);
+        return data;
+    }
+    
+    /**
+     * Enforce uniqueness among immediate child EQNodes.
+     *
+     * @param node
+     *            - an ASTOrNode or ASTAndNode
+     */
+    private void removeDuplicateTerms(JexlNode node) {
+        
+        // 1. Node must contain at least 2 EQ nodes to be evaluated
+        if (!hasMinimumRequiredEqNodes(node)) {
+            if (log.isDebugEnabled())
+                log.debug("Node must have 2 EQNodes to be evaluated.");
+            return;
+        }
+        
+        // 2. Remove duplicate EQ nodes from child nodes, thus enforcing term uniqueness for node.
+        List<JexlNode> uniqueChildren = removeDuplicateEqNodes(node);
+        
+        // 3. Check to see if any duplicate children were removed
+        if (node.jjtGetNumChildren() - uniqueChildren.size() == 0) {
+            if (log.isDebugEnabled())
+                log.debug("No duplicate child EQNodes found for current node.");
+            return;
+        }
+        
+        if (uniqueChildren.size() != 1) {
+            
+            // 4. With two or more child nodes remaining, simply replace the existing children.
+            children(node, uniqueChildren.toArray(new JexlNode[0]));
+            
+        } else {
+            
+            // 5. If only one child remains, replace the parent's child with the unique child.
+            // Pull up the unique child node such that
+            // PARENT -> CURRENT -> {EQ_Child_0, EQ_Child_0}
+            // PARENT -> EQ_Child_0
+            replaceChild(node.jjtGetParent(), node, uniqueChildren.get(0));
+            
+            // if (FOO == 'bar') unwrap it such that FOO == 'bar'
+            // call to unwrap will only execute if the node is wrapped in a reference expression (parens)
+            unwrap(node, uniqueChildren.get(0));
```

Comment:
There may be no need for flattening here, since we have a flattening visitor that is used elsewhere

Reply:
In a contrived case of (A & A) || (A & B & B), that will reduce to two single A terms that are wrapped in parens. If they aren't unwrapped then this visitor will have problems further reducing the query tree. 

This unwrapping shouldn't be necessary but it's there as a safeguard.